src/main/java/reactor/pool/introspection/SamplingAllocationStrategy.java

```diff
@@ -0,0 +1,234 @@
+	static AllocationStrategy sizeBetweenHelper(int min, int max) {
+		AllocationStrategy[] as = new AllocationStrategy[1];
+		PoolBuilder.from(Mono.empty())
+		           .sizeBetween(min, max)
+		           .build(config -> {
+			           as[0] = config.allocationStrategy();
+			           return null;
+		           });
+		return as[0];
+	}
+
+	/**
+	 * An {@link SamplingAllocationStrategy} that wraps a {@link PoolBuilder#sizeBetween(int, int) sizeBetween} {@link AllocationStrategy}
+	 * and samples calls to {@link AllocationStrategy#getPermits(int)} and {@link AllocationStrategy#returnPermits(int)}.
+	 *
+	 * @param min the minimum number of permits (see {@link PoolBuilder#sizeBetween(int, int)})
+	 * @param max the maximum number of permits (see {@link PoolBuilder#sizeBetween(int, int)})
+	 * @param getPermitsSamplingRate the sampling rate for {@link AllocationStrategy#getPermits(int)} calls (0.0d to 1.0d)
+	 * @param returnPermitsSamplingRate the sampling rate for {@link AllocationStrategy#returnPermits(int)} calls (0.0d to 1.0d)
+	 * @return a sampled {@link AllocationStrategy} that otherwise behaves like the {@link PoolBuilder#sizeBetween(int, int)} strategy
+	 */
+	public static AllocationStrategy sizeBetweenWithSampling(int min, int max, double getPermitsSamplingRate, double returnPermitsSamplingRate) {
+		return new SamplingAllocationStrategy(sizeBetweenHelper(min, max), getPermitsSamplingRate, returnPermitsSamplingRate);
+	}
+
+	/**
+	 * An {@link AllocationStrategy} that wraps a {@code delegate} and samples calls to {@link AllocationStrategy#getPermits(int)}
+	 * and {@link AllocationStrategy#returnPermits(int)}. Only a given percentage of calls are sampled, and these are stored as stacktraces
+	 * in the sampling strategy. If a {@link AllocationStrategy#returnPermits(int)} call fails, a new {@link IllegalArgumentException}
+	 * is thrown with a composite cause that represents all the sampled calls (see {@link Exceptions#unwrapMultiple(Throwable)}).
+	 * <p>
+	 * The two types of calls are sampled separately. The reservoir sampling algorithm is used to instantiate a sampling decision
+	 * bitset over 100 slots, which are then applied live to windows of 100 calls.
+	 *
+	 * @param delegate the underlying {@link AllocationStrategy}
+	 * @param getPermitsSamplingRate the sampling rate for {@link AllocationStrategy#getPermits(int)} calls (0.0d to 1.0d)
+	 * @param returnPermitsSamplingRate the sampling rate for {@link AllocationStrategy#returnPermits(int)} calls (0.0d to 1.0d)
+	 * @return a sampled {@link AllocationStrategy} that otherwise behaves like the {@code delegate}
+	 */
+	public static AllocationStrategy withSampling(AllocationStrategy delegate, double getPermitsSamplingRate, double returnPermitsSamplingRate) {
+		return new SamplingAllocationStrategy(delegate, getPermitsSamplingRate, returnPermitsSamplingRate);
+	}
+
+	final AllocationStrategy    delegate;
+
+	final LinkedList<Throwable> gettingSamples;
```

Comment:
maybe I could make these lists `public final` for further introspection, as well as the `delegate`, wdyt @violetagg ?

Reply:
yep or getter methods